File: developers/academy/py/starter_multimodal_data/_snippets/103_searches.py
# START-ANY
import weaviate
import weaviate.classes.query as wq
import os

# END-ANY

# FilteredSemanticSearch
from datetime import datetime

# END FilteredSemanticSearch

# START-ANY

# END-ANY

headers = {"X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")}
client = weaviate.connect_to_local(
    port=8280,
    grpc_port=50251,
    headers=headers
)

# START-ANY
# Instantiate your client (not shown). e.g.:
# headers = {"X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")}  # Replace with your OpenAI API key
# client = weaviate.connect_to_local(headers=headers)

# END-ANY


# MetadataMultimodalSearch

def url_to_base64(url):
    import requests
    import base64

    image_response = requests.get(url)
    content = image_response.content
    return base64.b64encode(content).decode("utf-8")


# Get the collection
movies = client.collections.get("MovieMM")

# Perform query
src_img_path = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
query_b64 = url_to_base64(src_img_path)

response = movies.query.near_image(
    near_image=query_b64,
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]  # To include the poster property in the response (`blob` properties are not returned by default)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END MetadataMultimodalSearch


print("\n\n")

client.connect()


# MetadataSemanticSearch
# Get the collection
movies = client.collections.get("MovieMM")

# Perform query
response = movies.query.near_text(
    query="red",
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]  # To include the poster property in the response (`blob` properties are not returned by default)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END MetadataSemanticSearch


print("\n\n")

client.connect()

# MetadataBM25Search
# Get the collection
movies = client.collections.get("MovieMM")

# Perform query
response = movies.query.bm25(
    query="history", limit=5, return_metadata=wq.MetadataQuery(score=True)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"BM25 score: {o.metadata.score:.3f}\n"
    )  # Print the BM25 score of the object from the query

client.close()
# END MetadataBM25Search


print("\n\n")

client.connect()

# MetadataHybridSearch
# Get the collection
movies = client.collections.get("MovieMM")

# Perform query
response = movies.query.hybrid(
    query="history", limit=5, return_metadata=wq.MetadataQuery(score=True)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Hybrid score: {o.metadata.score:.3f}\n"
    )  # Print the hybrid search score of the object from the query

client.close()
# END MetadataHybridSearch


print("\n\n")

client.connect()

# FilteredSemanticSearch
# Get the collection
movies = client.collections.get("MovieMM")

# Perform query
response = movies.query.near_text(
    query="dystopian future",
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    # highlight-start
    filters=wq.Filter.by_property("release_date").greater_than(datetime(2020, 1, 1))
    # highlight-end
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END FilteredSemanticSearch

File: developers/academy/js/starter_multimodal_data/_snippets/103_searches.ts
// START-ANY
import 'dotenv/config'
import weaviate, { WeaviateClient, type Filter } from "weaviate-client";
let client: WeaviateClient;

// END-ANY

client = await weaviate.connectToWeaviateCloud(
  process.env.WCD_URL as string,
  {
    authCredentials: new weaviate.ApiKey(process.env.WCD_API_KEY as string),
    headers: {
      'X-Cohere-Api-Key': process.env.COHERE_APIKEY as string,  // Replace with your inference API key
    }
  }
)

// START-ANY
// Instantiate your client (not shown). e.g.:
// headers = {"X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")}  # Replace with your OpenAI API key
// client = weaviate.connect_to_local(headers=headers)

// MetadataMultimodalSearch


async function urlToBase64(imageUrl: string) {
  const response = await fetch(imageUrl);
  const content = await response.buffer();
  return content.toString('base64');
}

// END-ANY

// SinglePromptGeneration
// Get the collection
const movies = client.collections.get("MovieMM")

// Perform query
const srcImgPath = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
const queryB64 = await urlToBase64(srcImgPath)

const response = await movies.query.nearImage(queryB64,
   {
  limit: 5,
  returnMetadata: ['distance'],
  returnProperties: ["title", "tmdb_id", "release_date"]
},
)

// Inspect the response
for (let item of response.objects) {
  console.log(`${item.properties.title} - ${item.properties.release_date}`)
   // Print the title and release year (note the release date is a datetime object)
   console.log(`Distance to query: ${item.metadata?.distance}`)
   // Print the distance of the object from the query


}
client.close()
// END MetadataMultimodalSearch

console.log("\n\n")


// MetadataSemanticSearch
// Get the collection
const movies = client.collections.get("MovieMM")

// Perform query
const srcImgPath = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
const queryB64 = await urlToBase64(srcImgPath)

const response = await movies.query.nearImage(queryB64,
   {
  limit: 5,
  returnMetadata: ['distance'],
  returnProperties: ["title", "tmdb_id", "release_date"]
},
)

// Inspect the response
for (let item of response.objects) {
  console.log(`${item.properties.title} - ${item.properties.release_date}`)
   // Print the title and release year (note the release date is a datetime object)
   console.log(`Distance to query: ${item.metadata?.distance}`)
   // Print the distance of the object from the query


}
client.close()
// END MetadataSemanticSearch

// MetadataBM25Search
// Get the collection
const movies = client.collections.get("MovieMM")

// Perform query
const srcImgPath = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
const queryB64 = await urlToBase64(srcImgPath)

const response = await movies.query.bm25("history",
   {
  limit: 5,
  returnMetadata: ['score'],
},
)

// Inspect the response
for (let item of response.objects) {
  console.log(`${item.properties.title} - ${item.properties.release_date}`)
   // Print the title and release year (note the release date is a datetime object)
   console.log(`BM25 score: ${item.metadata?.score}`)
   // Print the distance of the object from the query


}
client.close()
// END MetadataBM25Search


print("\n\n")

client.connect()

// MetadataHybridSearch
// Get the collection
const movies = client.collections.get("MovieMM")

// Perform query
const srcImgPath = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
const queryB64 = await urlToBase64(srcImgPath)

const response = await movies.query.hybrid("history",
   {
  limit: 5,
  returnMetadata: ['score'],
  returnProperties: ["title", "tmdb_id", "release_date"]
},
)

// Inspect the response
for (let item of response.objects) {
  console.log(`${item.properties.title} - ${item.properties.release_date}`)
   // Print the title and release year (note the release date is a datetime object)
   console.log(`Hybrid score: ${item.metadata?.score}`)
   // Print the hybrid score of the object from the query


}
client.close()
// END MetadataHybridSearch


print("\n\n")

client.connect()

// FilteredSemanticSearch
// Get the collection
const movies = client.collections.get("MovieMM")

// Perform query
const srcImgPath = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
const queryB64 = await urlToBase64(srcImgPath)
const filterTime = new Date(2020, 5, 10)

const response = await movies.query.nearText("dystopian future",
   {
  limit: 5,
  returnMetadata: ['distance'],
  filters: movies.filter.byProperty("release_date").greaterThan(filterTime)
},
)

// Inspect the response
for (let item of response.objects) {
  console.log(`${item.properties.title} - ${item.properties.release_date}`)
   // Print the title and release year (note the release date is a datetime object)
   console.log(`Distance to query: ${item.metadata?.distance}`)
   // Print the distance of the object from the query


}
client.close()
// END FilteredSemanticSearch